packages/task/src/node/custom/custom-task.ts

```diff
@@ -0,0 +1,73 @@
+/********************************************************************************
+ * Copyright (C) 2021 ByteDance and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, named } from 'inversify';
+import { ILogger, MaybePromise } from '@theia/core/lib/common/';
+import { Task, TaskOptions } from '../task';
+import { TaskManager } from '../task-manager';
+import { TaskInfo } from '../../common/task-protocol';
+import { Process } from '@theia/process/lib/node';
+
+export const TaskCustomOptions = Symbol('TaskCustomOptions');
+export interface TaskCustomOptions extends TaskOptions {
+    process: Process
+}
+
+export const TaskFactory = Symbol('TaskFactory');
+export type TaskFactory = (options: TaskCustomOptions) => CustomTask;
+
+/** Represents a Task launched as a fake process by `CustomTaskRunner`. */
+@injectable()
+export class CustomTask extends Task {
+
+    constructor(
+        @inject(TaskManager) protected readonly taskManager: TaskManager,
+        @inject(ILogger) @named('task') protected readonly logger: ILogger,
+        @inject(TaskCustomOptions) protected readonly options: TaskCustomOptions
+    ) {
+        super(taskManager, logger, options);
+        this.logger.info(`Created new custom task, id: ${this.id}, context: ${this.context}`);
+    }
+
+    kill(): Promise<void> {
+        return Promise.resolve();
+    }
+
+    getRuntimeInfo(): MaybePromise<TaskInfo> {
+        return {
+            taskId: this.id,
+            ctx: this.context,
+            config: this.options.config,
+            terminalId: this.process.id,
+            processId: this.process.id,
+        };
+    }
+
+    public callbackTaskComplete(exitCode: number | undefined): MaybePromise<void> {
+        this.fireTaskExited({
+            taskId: this.taskId,
+            ctx: this.context,
+            config: this.options.config,
+            terminalId: this.process.id,
+            processId: this.process.id,
+            code: exitCode,
```

Comment:
I think this should be:
```ts
  code: exitCode || 0,
```

VS Code works a bit differently, nothing happens on the callback:
https://github.com/microsoft/vscode/blob/d1e5c8cddf4fba06aad2e4cffd5b146f2de10700/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L372-L375

and we want to avoid logging an error on custom task execution here:
https://github.com/eclipse-theia/theia/blob/be686eef3e394bc521480fe8203aa7467e2cb98b/packages/task/src/browser/task-service.ts#L313

What do you think?

Reply:
I think so too!